_posts/2023-11-22-privacy-preserving-scikit-learn.md

```diff
@@ -0,0 +1,182 @@
+
+FHE is a technology that enables application providers to build cloud-based applications that preserve user privacy and [Concrete ML](https://github.com/zama-ai/concrete-ml) is a machine learning toolkit that converts models to use FHE. Concrete ML leverages the powerful and robust model training algorithms in scikit-learn to train FHE compatible models without requiring any knowledge of cryptography.
+
+Concrete ML uses scikit-learn as a basis for building FHE compatible models due to scikit-learn’s excellent ease of use, extensibility, robustness and wide palette of tools for building, validating and tuning data pipelines. While deep learning is performant on unstructured data, it often requires hyper-parameter tuning to achieve high accuracy. On many use cases, especially on structured data, scikit-learn excels through the robustness of its training algorithms.
+
+## Training a model locally and deploying it securely
+
+When all training data is available to the data scientist, training is secure as no data leaves their machine and only inference needs to be secured when the model is deployed. However, training models for FHE secured inference imposes some constraints on model training. While in the past using FHE required cryptographic expertise, tools like Concrete ML abstract away the cryptography and make FHE accessible to data scientists. Furthermore, FHE adds computation overhead which means that machine learning models may need to be tuned for both accuracy and runtime latency. Concrete ML makes such tuning easy by leveraging parameter search using scikit-learn utility classes such as `GridSearchCV`.
+
+To use Concrete ML to train a model locally the syntax is the same as for scikit-learn. Explanations can also be found in this [video tutorial](https://www.youtube.com/watch?v=p5Y25FWLrH0). For a logistic regression model on MNIST simply run the following snippets:
+
+```
+from sklearn.datasets import fetch_openml
+from sklearn.model_selection import train_test_split
+
+mnist_dataset = fetch_openml("mnist_784")
+
+x_train, x_test, y_train, y_test = train_test_split(
+    mnist_dataset.data,
+    mnist_dataset.target.astype("int"),
+    test_size=10000,
+)
+```
+
+Next, fit the Concrete ML logistic regression model which is a drop-in replacement of scikit-learn’s equivalent. An additional step, compilation, is necessary to produce an FHE computation circuit that performs the inference on encrypted data. Compilation, which is done by [Concrete](https://github.com/zama-ai/concrete), is the process of turning a program into its FHE equivalent, working directly over encrypted data.
+
+
+```
+from concrete.ml.sklearn.linear_model import LogisticRegression
+
+model = LogisticRegression(penalty="l2")
+model.fit(X=x_train, y=y_train)
+model.compile(x_train)
```

Comment:
I'm curious about the internals of this `.compile` step? Could you share a diagram with a brief explainer of what might go wrong if one doesn't run that line?

Reply:
Is it sufficient now (with in particular an image)? Else, we can link to https://docs.zama.ai/concrete-ml/advanced-topics/compilation#compilation-to-fhe, but it's more complicated and maybe a bit out of scope for this blog: your choice, you tell me